File: mailferry/report.py
"""
MailFerry - IMAP Migration & Sync
High-Performance Native IMAP Migration Engine

Copyright (C) 2026 Andy Saputra <andy@saputra.org>

https://saputra.org
https://github.com/ajsap/mailferry

Licensed under the GNU Affero General Public License v3.0 (AGPL-3.0).
This program is free software: you can redistribute it and/or modify it
under the terms of the GNU Affero General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

Contributions welcome: submit issues, feature requests and pull requests
at https://github.com/ajsap/mailferry

Session log, per-mailbox logs, results.csv, end-of-run summary.
"""
from __future__ import annotations

import csv
import json
import threading
import time
from collections import Counter
from datetime import datetime
from pathlib import Path
from typing import Optional

from . import SLOGAN, banner_line
from .progress.dashboard import c
from .util import fmt_bytes, fmt_dhms, pct, safe_name


class Session:
    """Timestamped, thread-safe master log (appends across runs)."""

    def __init__(self, path, json_path: str = ""):
        self.path = Path(path)
        self.json_path = Path(json_path) if json_path else None
        self._lock = threading.Lock()
        self.hub = None                      # set by the CLI; feeds the Logs view

    def log(self, msg: str):
        with self._lock:
            ts = datetime.now().isoformat(timespec="seconds")
            try:
                with open(self.path, "a", encoding="utf-8") as f:
                    f.write(f"{ts} {msg}\n")
            except OSError:
                pass
            if self.hub is not None:
                sev = "ERROR" if "ERROR" in msg else ("WARN" if "fail" in msg.lower() else "INFO")
                self.hub.log_event(sev, "-", msg)
            if self.json_path:
                try:
                    with open(self.json_path, "a", encoding="utf-8") as f:
                        f.write(json.dumps({"ts": ts, "event": msg}) + "\n")
                except OSError:
                    pass


class MailboxLogger:
    """Per-mailbox operational log. Never raises."""

    def __init__(self, path, title: str):
        self.path = Path(path)
        self._lock = threading.Lock()
        self.log(f"# {banner_line()}")
        self.log(f"# {SLOGAN}")
        self.log(f"# Migration log — {title}")

    def log(self, msg: str):
        with self._lock:
            try:
                with open(self.path, "a", encoding="utf-8") as f:
                    f.write(f"{datetime.now().isoformat(timespec='seconds')} {msg}\n")
            except OSError:
                pass


def logger_factory(logs_dir):
    def factory(spec):
        path = Path(logs_dir) / f"{spec.index:03d}_{safe_name(spec.src.user)}.log"
        return MailboxLogger(path, f"{spec.src.label} -> {spec.dst.label}")
    return factory


EXIT_OF = {"SUCCESS": "0", "WARNINGS": "0", "PARTIAL": "1", "FAILED": "1", "STALE": "1"}


def write_results_csv(snap, logs_dir) -> Path:
    path = Path(logs_dir) / "results.csv"
    with open(path, "w", newline="", encoding="utf-8") as f:
        w = csv.writer(f)
        w.writerow(["index", "olduser", "newuser", "status", "exit_code", "elapsed_seconds",
                    "log_file", "notes", "error_type", "attempts",
                    "msgs_new", "msgs_adopted", "msgs_skipped", "msgs_failed", "bytes_done",
                    "folders", "reconnects", "retries"])
        for m in snap["mailboxes"]:
            elapsed = int((m["end"] or time.time()) - m["start"]) if m["start"] else 0
            notes = ""
            if m["status"] == "CANCELLED":
                notes = "interrupted before completion — will be retried on the next run"
            elif m["status"] == "SKIPPED":
                notes = "already completed in a previous run — skipped (--skip-completed)"
            elif m["status"] == "PARTIAL":
                notes = "completed with skipped messages or failed folders — re-run to retry the gaps"
            elif m["status"] == "WARNINGS":
                nf = m.get("failed", 0) + m["skipped"]
                notes = (f"completed with warnings — {nf} message(s) could not be migrated "
                         "(Failed Message Registry: mailferry failed / retry-failed)")
            elif m["status"] == "STALE":
                notes = ("no progress and automatic recovery failed — verify the server, "
                         "then re-run (resumes from the last confirmed state)")
            elif m["status"] == "REMOTE":
                notes = ("being processed by another cluster worker"
                         + (f" ({m['detail'].replace('processed by ', '')})"
                            if m.get("detail") else ""))
            elif m["status"] == "FAILED":
                notes = m["error"][:300]
            w.writerow([m["index"], m["label"], m["label2"], m["status"], EXIT_OF.get(m["status"], ""),
                        elapsed, m["log"], notes, m["error"][:120],
                        m["attempt"] if m["start"] else "",
                        m["appended"], m["adopted"], m["skipped"], m.get("failed", 0),
                        m["bytes_done"],
                        f"{m['fi']}/{m['ft']}" if m["ft"] else "",
                        m["src"]["reconnects"] + m["dst"]["reconnects"], m["retries"]])
    return path


def print_summary(snap, results_path, cfg, runtime_secs: float, interrupted: bool):
    agg, counts = snap["agg"], snap["counts"]
    mbs = snap["mailboxes"]
    ok = [m for m in mbs if m["status"] == "SUCCESS"]
    warned = [m for m in mbs if m["status"] == "WARNINGS"]
    partial = [m for m in mbs if m["status"] == "PARTIAL"]
    failed = [m for m in mbs if m["status"] == "FAILED"]
    stale = [m for m in mbs if m["status"] == "STALE"]
    cancelled = [m for m in mbs if m["status"] == "CANCELLED"]
    attempted = len(ok) + len(warned) + len(partial) + len(failed) + len(stale)

    def kv(label, value):
        print(f"{label:<21}: {value}")

    print()
    title = (f"{banner_line()} — run interrupted, partial results" if interrupted
             else f"{banner_line()} — run complete")
    print(c(title, "yellow" if interrupted else "bold"))
    kv("Total runtime", fmt_dhms(runtime_secs))
    kv("Mailboxes in CSV", len(mbs))
    if snap["skipped_prior"]:
        kv("Skipped (done)", f"{snap['skipped_prior']} — previously completed (--skip-completed)")
    kv("Attempted this run", attempted)
    kv("Successful", c(str(len(ok)), "green"))
    if warned:
        kv("With warnings", c(f"{len(warned)} — completed; a few messages are in the "
                              "Failed Message Registry", "yellow"))
    if partial:
        kv("Partial", c(str(len(partial)), "yellow"))
    kv("Failed", c(str(len(failed)), "red") if failed else "0")
    if stale:
        kv("Stale", c(f"{len(stale)} — automatic recovery failed; see below", "red"))
    remote = [m for m in mbs if m["status"] == "REMOTE"]
    if remote:
        kv("On other workers", c(f"{len(remote)} — processed by other cluster "
                                 "instances sharing this State Database", "cyan"))
    if cancelled:
        kv("Cancelled", c(str(len(cancelled)), "yellow"))
    if attempted:
        kv("Success rate", f"{(len(ok) + len(warned)) * 100 / attempted:.0f}% (of attempted)")
    kv("Messages synced", f"{agg['msgs_done']:,} of {agg['msgs_total']:,} "
                          f"({pct(agg['msgs_done'], agg['msgs_total'])})")
    kv("  copied (new)", f"{agg['appended']:,}")
    if agg["adopted"]:
        kv("  adopted (dup-safe)", f"{agg['adopted']:,} — already on destination, not re-copied")
    if agg["skipped_msgs"]:
        kv("  skipped msgs", c(f"{agg['skipped_msgs']:,} (see per-mailbox logs)", "yellow"))
    if agg.get("failed_msgs"):
        kv("  failed msgs", c(f"{agg['failed_msgs']:,} — Failed Message Registry "
                              "(mailferry failed / retry-failed)", "red"))
    kv("Data synced", fmt_bytes(agg["bytes_done"]))
    kv("Wire traffic", f"down {fmt_bytes(agg['wire_rx'])} / up {fmt_bytes(agg['wire_tx'])}")
    if runtime_secs > 0 and agg["wire_tx"]:
        kv("Avg throughput", f"{fmt_bytes(agg['wire_tx'] / runtime_secs)}/s (upload wire)")
    done_timed = [m for m in ok if m["start"] and m["end"]]
    if done_timed:
        durs = [(m["end"] - m["start"], m) for m in done_timed]
        kv("Avg mailbox time", fmt_dhms(sum(d for d, _ in durs) / len(durs)))
        slow = max(durs, key=lambda t: t[0])
        fast = min(durs, key=lambda t: t[0])
        kv("Slowest mailbox", f"{slow[1]['label']} ({fmt_dhms(slow[0])})")
        kv("Fastest mailbox", f"{fast[1]['label']} ({fmt_dhms(fast[0])})")
    if agg["reconnects"]:
        kv("Reconnects", agg["reconnects"])
    if agg["retries"]:
        kv("Retries used", agg["retries"])
    st = snap.get("stalls") or {}
    if st.get("detected"):
        kv("Stalls detected", f"{st['detected']} — auto-recovered {st.get('recovered', 0)}"
           + (c(f" · recovery failed {st['failed']}", "red") if st.get("failed") else ""))
    kv("Per-mailbox logs", cfg.logs_dir)
    kv("Session log", str(Path(cfg.logs_dir) / "session.log"))
    kv("Results CSV", str(results_path))
    kv("State Database", "ephemeral (--ephemeral) — nothing persisted" if cfg.ephemeral else cfg.db_path)

    for m in warned:
        nf = m.get("failed", 0) + m["skipped"]
        tot = m["msgs_total"] or (m["msgs_done"] + nf)
        pw = (m["msgs_done"] * 100.0 / tot) if tot else 100.0
        print(c(f"\n{m['label']} — COMPLETED WITH WARNINGS", "yellow"))
        kv("  Total messages", f"{tot:,}")
        kv("  Migrated", f"{m['msgs_done']:,}")
        kv("  Failed", c(str(nf), "red"))
        kv("  Success rate", f"{pw:.2f}%")

    problems = failed + partial + stale
    if problems or cancelled:
        reasons = Counter((m["error"] or "unknown error") for m in problems if m["error"])
        if reasons:
            print()
            print(c("Error summary:", "red"))
            for reason, n in reasons.most_common():
                print(f"  {n} × {reason[:140]}")
        if failed:
            print(c("Failed mailboxes:", "red"))
            for m in failed:
                att = f" (after {m['attempt']} attempts)" if m["attempt"] > 1 else ""
                print(f"  - {m['label']} — {m['error'][:140]}{att}  (log: {m['log'] or '-'})")
        if stale:
            print(c("Stale mailboxes (no progress; automatic recovery failed):", "red"))
            for m in stale:
                print(f"  - {m['label']} — {m['error'][:140] or 'no progress'}  "
                      f"(log: {m['log'] or '-'})")
            print(c("  Verify the server is reachable, then re-run — MailFerry resumes "
                    "from the last confirmed state (never duplicates).", "cyan"))
        if partial:
            print(c("Partial mailboxes (gaps will be retried next run):", "yellow"))
            for m in partial:
                print(f"  - {m['label']} — skipped {m['skipped']}, error: {m['error'][:120] or '-'}")
        if cancelled:
            print(c(f"Cancelled mailboxes ({len(cancelled)}):", "yellow"))
            for m in cancelled:
                print(f"  - {m['label']}")
        if not cfg.ephemeral:
            print()
            print(c("Resume: re-run the same command — completed messages are never "
                    "re-copied (per-UID state + fingerprint adoption).", "cyan"))


def write_failed_report(rows, logs_dir) -> Optional[Path]:
    """Export the outstanding Failed Message Registry beside results.csv."""
    if not rows:
        return None
    path = Path(logs_dir) / "failed_messages.csv"
    with open(path, "w", newline="", encoding="utf-8") as f:
        w = csv.writer(f)
        w.writerow(["mailbox", "folder", "uid", "message_id", "subject", "sender",
                    "date", "size", "failure_type", "reason", "fail_count",
                    "first_failure", "last_failure", "status"])
        for r in rows:
            w.writerow([r.get("mailbox", ""), r["folder"], r["src_uid"],
                        r["message_id"], r["subject"], r["sender"], r["date"],
                        r["size"], r["ftype"], r["reason"], r["fail_count"],
                        datetime.fromtimestamp(r["first_ts"]).isoformat(timespec="seconds")
                        if r["first_ts"] else "",
                        datetime.fromtimestamp(r["last_ts"]).isoformat(timespec="seconds")
                        if r["last_ts"] else "",
                        r["status"]])
    return path


def print_failed_section(rows, path):
    """Dedicated Failed Messages section of the final report."""
    if not rows:
        return
    print()
    print(c(f"Failed Messages ({len(rows)} outstanding):", "red"))
    for r in rows[:20]:
        subj = (r["subject"] or "-")[:48]
        print(f"  - {r.get('mailbox', '?')} · {r['folder']} · UID {r['src_uid']} · "
              f"{fmt_bytes(r['size'])} · {r['ftype']}")
        print(f"      {subj} · {(r['sender'] or '-')[:40]} · {(r['date'] or '-')[:26]}")
        print(f"      reason: {(r['reason'] or '-')[:110]}")
    if len(rows) > 20:
        print(f"  … +{len(rows) - 20} more")
    if path:
        print(c(f"  Full report: {path}   (JSON: mailferry failed --json)", "cyan"))
    print(c("  Retry them any time: mailferry retry-failed CSV — circumstances "
            "change (quota, upgrades, filters).", "cyan"))

File: internal/report/report.go
// MailFerry — IMAP Migration & Sync
// High-Performance Native IMAP Migration Engine
//
// Copyright (C) 2026 Andy Saputra
// Author: Andy Saputra <andy@saputra.org>
// SPDX-License-Identifier: AGPL-3.0-or-later
//
// This file is part of MailFerry (https://github.com/ajsap/mailferry).
// Licensed under the GNU Affero General Public License v3.0 or later;
// see the LICENSE file for details.

// Package report: session log, per-mailbox logs, results.csv, summary.
package report

import (
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"encoding/json"

	"github.com/ajsap/mailferry/v2/internal/config"
	"github.com/ajsap/mailferry/v2/internal/engine"
	"github.com/ajsap/mailferry/v2/internal/identity"
	"github.com/ajsap/mailferry/v2/internal/state"
	"github.com/ajsap/mailferry/v2/internal/util"
)

// Session: timestamped, goroutine-safe master log. An optional NDJSON
// event log (--json-logs) receives {"ts","event"} rows in parallel.
type Session struct {
	mu       sync.Mutex
	path     string
	jsonPath string
}

func NewSession(path string, jsonPath ...string) *Session {
	s := &Session{path: path}
	if len(jsonPath) > 0 {
		s.jsonPath = jsonPath[0]
	}
	return s
}

func (s *Session) Log(msg string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	ts := util.NowISO()
	f, err := os.OpenFile(s.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
	if err == nil {
		fmt.Fprintf(f, "%s %s\n", ts, msg)
		f.Close()
	}
	if s.jsonPath != "" {
		if jf, err := os.OpenFile(s.jsonPath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600); err == nil {
			enc, _ := json.Marshal(map[string]string{"ts": ts, "event": msg})
			jf.Write(append(enc, '\n'))
			jf.Close()
		}
	}
}

// StartProgressNDJSON writes one JSON snapshot per second (--json-progress)
// until the returned stop function is called. Purely observational.
func StartProgressNDJSON(path string, stats *engine.Stats) (stop func()) {
	done := make(chan struct{})
	go func() {
		t := time.NewTicker(1 * time.Second)
		defer t.Stop()
		for {
			select {
			case <-done:
				return
			case <-t.C:
				snap := stats.Snapshot()
				agg := snap.Agg()
				row := map[string]any{
					"ts": util.NowISO(), "counts": agg.Counts,
					"msgs_done": agg.MsgsDone, "msgs_total": agg.MsgsTotal,
					"bytes_done": agg.BytesDone, "bytes_total": agg.BytesTotal,
					"wire_rx": agg.WireRX, "wire_tx": agg.WireTX,
					"failed_msgs": agg.FailedMsgs, "reconnects": agg.Reconnects,
				}
				if f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600); err == nil {
					enc, _ := json.Marshal(row)
					f.Write(append(enc, '\n'))
					f.Close()
				}
			}
		}
	}()
	return func() { close(done) }
}

// MailboxLoggerFactory: per-mailbox operational logs, never raises.
func MailboxLoggerFactory(logsDir string) engine.LoggerFactory {
	return func(spec config.MailboxSpec) func(string, ...any) {
		path := filepath.Join(logsDir,
			fmt.Sprintf("%03d_%s.log", spec.Index, util.SafeName(spec.Src.User)))
		var mu sync.Mutex
		write := func(line string) {
			mu.Lock()
			defer mu.Unlock()
			f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
			if err != nil {
				return
			}
			defer f.Close()
			fmt.Fprintf(f, "%s %s\n", util.NowISO(), line)
		}
		write("# " + identity.BannerLine())
		write("# " + identity.Slogan)
		write(fmt.Sprintf("# Migration log — %s -> %s", spec.Src.Label(), spec.Dst.Label()))
		return func(format string, a ...any) { write(fmt.Sprintf(format, a...)) }
	}
}

var exitOf = map[string]string{"SUCCESS": "0", "WARNINGS": "0", "PARTIAL": "1",
	"FAILED": "1", "STALE": "1"}

// WriteResultsCSV writes logs/results.csv (same columns as v1.x).
func WriteResultsCSV(snap engine.Snapshot, logsDir string) string {
	path := filepath.Join(logsDir, "results.csv")
	f, err := os.Create(path)
	if err != nil {
		return path
	}
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()
	w.Write([]string{"index", "olduser", "newuser", "status", "exit_code",
		"elapsed_seconds", "log_file", "notes", "error_type", "attempts",
		"msgs_new", "msgs_adopted", "msgs_skipped", "msgs_failed", "bytes_done",
		"folders", "reconnects", "retries"})
	for _, m := range snap.Mailboxes {
		elapsed := int64(0)
		if !m.Start.IsZero() {
			end := m.End
			if end.IsZero() {
				end = time.Now()
			}
			elapsed = int64(end.Sub(m.Start).Seconds())
		}
		notes := ""
		switch m.Status {
		case "CANCELLED":
			notes = "interrupted before completion — will be retried on the next run"
		case "SKIPPED":
			notes = "already completed in a previous run — skipped (--skip-completed)"
		case "PARTIAL":
			notes = "completed with skipped messages or failed folders — re-run to retry the gaps"
		case "WARNINGS":
			notes = fmt.Sprintf("completed with warnings — %d message(s) could not be migrated",
				m.Skipped+m.FailedMsgs)
		case "FAILED":
			notes = trunc(m.Error, 300)
		}
		folders := ""
		if m.FoldersTotal > 0 {
			folders = fmt.Sprintf("%d/%d", m.FolderIndex, m.FoldersTotal)
		}
		w.Write([]string{
			fmt.Sprint(m.Index), m.Label, m.Label2, m.Status, exitOf[m.Status],
			fmt.Sprint(elapsed), m.LogPath, notes, trunc(m.Error, 120),
			fmt.Sprint(m.Attempt), fmt.Sprint(m.Appended), fmt.Sprint(m.Adopted),
			fmt.Sprint(m.Skipped), fmt.Sprint(m.FailedMsgs), fmt.Sprint(m.BytesDone),
			folders, fmt.Sprint(m.Src.Reconnects + m.Dst.Reconnects), fmt.Sprint(m.Retries)})
	}
	return path
}

// PrintSummary: the end-of-run report.
func PrintSummary(snap engine.Snapshot, resultsPath string, cfg *config.Run,
	runtime float64, interrupted bool, colf func(string, string) string) {
	agg := snap.Agg()
	c := colf
	fmt.Println()
	title := identity.BannerLine() + " — run complete"
	if interrupted {
		title = identity.BannerLine() + " — run interrupted, partial results"
	}
	fmt.Println(c(title, "bold"))
	kv := func(k, v string) { fmt.Printf("%-21s: %s\n", k, v) }
	kv("Total runtime", util.FmtDHMS(runtime))
	kv("Mailboxes in CSV", fmt.Sprint(len(snap.Mailboxes)))
	attempted := agg.Counts["SUCCESS"] + agg.Counts["WARNINGS"] + agg.Counts["PARTIAL"] +
		agg.Counts["FAILED"] + agg.Counts["STALE"]
	kv("Attempted this run", fmt.Sprint(attempted))
	kv("Successful", c(fmt.Sprint(agg.Counts["SUCCESS"]), "green"))
	if n := agg.Counts["WARNINGS"]; n > 0 {
		kv("With warnings", c(fmt.Sprintf("%d — a few messages could not be migrated", n), "yellow"))
	}
	if n := agg.Counts["PARTIAL"]; n > 0 {
		kv("Partial", c(fmt.Sprint(n), "yellow"))
	}
	kv("Failed", fmt.Sprint(agg.Counts["FAILED"]))
	if n := agg.Counts["CANCELLED"]; n > 0 {
		kv("Cancelled", c(fmt.Sprint(n), "yellow"))
	}
	if attempted > 0 {
		kv("Success rate", fmt.Sprintf("%.0f%% (of attempted)",
			float64(agg.Counts["SUCCESS"]+agg.Counts["WARNINGS"])*100/float64(attempted)))
	}
	kv("Messages synced", fmt.Sprintf("%d of %d (%s)", agg.MsgsDone, agg.MsgsTotal,
		util.Pct(agg.MsgsDone, agg.MsgsTotal)))
	kv("  copied (new)", fmt.Sprint(agg.Appended))
	if agg.Adopted > 0 {
		kv("  adopted (dup-safe)", fmt.Sprintf("%d — already on destination, not re-copied", agg.Adopted))
	}
	if agg.SkippedMsgs > 0 {
		kv("  skipped msgs", c(fmt.Sprintf("%d (see per-mailbox logs)", agg.SkippedMsgs), "yellow"))
	}
	kv("Data synced", util.FmtBytes(float64(agg.BytesDone)))
	kv("Wire traffic", fmt.Sprintf("down %s / up %s",
		util.FmtBytes(float64(agg.WireRX)), util.FmtBytes(float64(agg.WireTX))))
	if runtime > 0 && agg.WireTX > 0 {
		kv("Avg throughput", util.FmtBytes(float64(agg.WireTX)/runtime)+"/s (upload wire)")
	}
	if agg.Reconnects > 0 {
		kv("Reconnects", fmt.Sprint(agg.Reconnects))
	}
	kv("Per-mailbox logs", cfg.LogsDir)
	kv("Session log", filepath.Join(cfg.LogsDir, "session.log"))
	kv("Results CSV", resultsPath)
	kv("State Database", cfg.DBPath)

	var failed, partial []engine.MBValues
	for _, m := range snap.Mailboxes {
		switch m.Status {
		case "FAILED":
			failed = append(failed, m)
		case "PARTIAL":
			partial = append(partial, m)
		}
	}
	if len(failed) > 0 {
		fmt.Println(c("Failed mailboxes:", "red"))
		for _, m := range failed {
			fmt.Printf("  - %s — %s\n", m.Label, trunc(m.Error, 140))
		}
	}
	if len(partial) > 0 {
		fmt.Println(c("Partial mailboxes (gaps will be retried next run):", "yellow"))
		for _, m := range partial {
			fmt.Printf("  - %s — skipped %d, error: %s\n", m.Label, m.Skipped, trunc(m.Error, 120))
		}
	}
	if len(failed)+len(partial) > 0 && !cfg.Ephemeral {
		fmt.Println(c("Resume: re-run the same command — completed messages are never re-copied.", "cyan"))
	}
}

func trunc(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

var _ = strings.TrimSpace

// PrintFailedSection renders the dedicated Failed Messages block and writes
// logs/failed_messages.csv when there are outstanding failures.
func PrintFailedSection(rows []state.FailedRow, outstanding int64, logsDir string,
	colf func(string, string) string) {
	if len(rows) == 0 {
		return
	}
	path := filepath.Join(logsDir, "failed_messages.csv")
	_ = WriteFailedCSVTo(rows, path)
	fmt.Println()
	fmt.Println(colf(fmt.Sprintf("Failed Messages (%d outstanding):", outstanding), "red"))
	limit := len(rows)
	if limit > 20 {
		limit = 20
	}
	for _, r := range rows[:limit] {
		fmt.Printf("  - %s · %s · UID %d · %s · %s\n", orNA(r.Mailbox), r.Folder,
			r.SrcUID, util.FmtBytes(float64(r.Size)), r.FType)
		fmt.Printf("      %s · %s\n", clipR(orNA(r.Subject), 48), clipR(r.Reason, 90))
	}
	if len(rows) > 20 {
		fmt.Printf("  … +%d more\n", len(rows)-20)
	}
	fmt.Println(colf("  Full report: "+path+"   (JSON: mailferry failed --json)", "cyan"))
	fmt.Println(colf("  Retry any time: mailferry retry-failed — circumstances change "+
		"(quota, upgrades, filters).", "cyan"))
}

func orNA(s string) string {
	if s == "" {
		return "?"
	}
	return s
}
func clipR(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

// WriteFailedCSVTo exports the Failed Message Registry to a CSV file.
func WriteFailedCSVTo(rows []state.FailedRow, path string) error {
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()
	w.Write([]string{"mailbox", "folder", "uid", "message_id", "subject", "sender",
		"date", "size", "failure_type", "reason", "fail_count", "first_failure",
		"last_failure", "status"})
	for _, r := range rows {
		w.Write([]string{r.Mailbox, r.Folder, fmt.Sprint(r.SrcUID), r.MessageID,
			r.Subject, r.Sender, r.Date, fmt.Sprint(r.Size), r.FType, r.Reason,
			fmt.Sprint(r.FailCount), tsISO(r.FirstTS), tsISO(r.LastTS), r.Status})
	}
	return nil
}

// FailedJSON prints the registry as JSON to stdout.
func FailedJSON(rows []state.FailedRow) int {
	type outRow struct {
		Mailbox, Folder, MessageID, Subject, Sender, Date, FType, Reason, Status string
		UID                                                                      uint32
		Size, FailCount                                                          int64
	}
	var out []outRow
	for _, r := range rows {
		out = append(out, outRow{r.Mailbox, r.Folder, r.MessageID, r.Subject, r.Sender,
			r.Date, r.FType, r.Reason, r.Status, r.SrcUID, r.Size, r.FailCount})
	}
	b, _ := json.MarshalIndent(out, "", "  ")
	fmt.Println(string(b))
	return 0
}

func tsISO(ts float64) string {
	if ts <= 0 {
		return ""
	}
	return time.Unix(int64(ts), 0).Format("2006-01-02T15:04:05")
}
